File: sdk/reproit-tui-ts/causal.ts
type AnyMap = Record<string, any>;
const SECRET = /password|passwd|secret|token|authorization|cookie|email|phone|api[-_. ]?key|publishable[-_. ]?key|private[-_. ]?key|access[-_. ]?key|signing[-_. ]?key/i;

export function redactCausal(value: any): any {
  if (Array.isArray(value)) return value.map(redactCausal);
  if (value && typeof value === "object") {
    const out: AnyMap = {};
    for (const key of Object.keys(value).sort()) {
      const child = value[key];
      out[key] = SECRET.test(key)
        ? `<reproit:${typeof child === "string" ? `string:length=${[...child].length}` : typeof child}>`
        : redactCausal(child);
    }
    return out;
  }
  return value;
}

function canonical(raw: string): string {
  try {
    const u = new URL(raw);
    const pairs = [...u.searchParams.entries()].sort(([ak, av], [bk, bv]) => ak.localeCompare(bk) || av.localeCompare(bv));
    u.search = "";
    for (const [k, v] of pairs) u.searchParams.append(k, v);
    return u.toString();
  } catch { return raw; }
}

/** Automatic global-fetch adapter for Node-based TUIs. It uses runner-owned
 * files rather than stdout/stderr, because both streams are the rendered PTY. */
export function installCausalFetch(excludePrefix?: string | null): () => void {
  const proc: any = (globalThis as any).process;
  const fs = proc?.getBuiltinModule?.("node:fs");
  const networkFile = proc?.env?.REPROIT_NETWORK_FILE;
  const actionFile = proc?.env?.REPROIT_ACTION_FILE;
  const capsulePath = proc?.env?.REPROIT_CAPSULE;
  const capabilitiesFile = proc?.env?.REPROIT_CAPABILITIES_FILE;
  const original: any = (globalThis as any).fetch;
  if (!fs || typeof original !== "function" || (!networkFile && !capsulePath)) return () => {};
  let capsule: AnyMap | null = null;
  try { if (capsulePath) capsule = JSON.parse(fs.readFileSync(capsulePath, "utf8")); } catch {}
  const exchanges: AnyMap[] = capsule?.exchanges || [];
  const used = new Set<number>();
  let ordinal = 0;
  let previous = 0;
  const actionIndex = () => {
    try { return Number(fs.readFileSync(actionFile, "utf8")) || 0; } catch { return 0; }
  };
  const mergeCapabilities = () => {
    try {
      const current = JSON.parse(fs.readFileSync(capabilitiesFile, "utf8"));
      current.http = { status: "captured", detail: "Node global fetch" };
      current.http_replay = capsule
        ? { status: "captured", detail: "Node global fetch fail-closed replay" }
        : { status: "captured", detail: "adapter supports capsule replay" };
      fs.writeFileSync(capabilitiesFile, JSON.stringify(current), { mode: 0o600 });
    } catch {}
  };
  mergeCapabilities();
  (globalThis as any).fetch = async (input: any, init: AnyMap = {}) => {
    const url = typeof input === "string" ? input : input?.url || String(input);
    if (excludePrefix && url.startsWith(excludePrefix)) return original(input, init);
    const action = actionIndex();
    if (action !== previous) { previous = action; ordinal = 0; }
    const thisOrdinal = ordinal++;
    const method = String(init.method || input?.method || "GET").toUpperCase();
    if (capsule) {
      const idx = exchanges.findIndex((e, i) => !used.has(i) && e.required && e.actor === (proc.env.REPROIT_DEVICE || "a") &&
        e.actionIndex === action && String(e.method).toUpperCase() === method && canonical(String(e.url)) === canonical(url));
      if (idx < 0) throw new Error(`CAPSULE:MISS ${method} ${url} action=${action}`);
      used.add(idx);
      const e = exchanges[idx];
      return new Response(typeof e.responseBody === "string" ? e.responseBody : JSON.stringify(e.responseBody ?? ""), {
        status: e.status, headers: e.responseHeaders,
      });
    }
    const response = await original(input, init);
    let responseBody: any;
    try {
      responseBody = /json/i.test(response.headers.get("content-type") || "")
        ? redactCausal(await response.clone().json()) : "<reproit:body>";
    } catch { responseBody = "<reproit:invalid-json>"; }
    let requestBody: any;
    if (typeof init.body === "string") {
      try { requestBody = redactCausal(JSON.parse(init.body)); }
      catch { requestBody = `<reproit:body:length=${init.body.length}>`; }
    }
    const headers: AnyMap = {};
    new Headers(init.headers || {}).forEach((v, k) => { headers[k] = SECRET.test(k) ? "<reproit:secret>" : v; });
    const responseHeaders: AnyMap = {};
    response.headers.forEach((v: string, k: string) => { responseHeaders[k] = SECRET.test(k) ? "<reproit:secret>" : v; });
    const actor = proc.env.REPROIT_DEVICE || "a";
    const exchange = { id: `${actor}-${action}-${thisOrdinal}`, actor, actionIndex: action, ordinal: thisOrdinal,
      protocol: new URL(url).protocol.replace(":", ""), method, url: canonical(url), requestHeaders: headers,
      ...(requestBody === undefined ? {} : { requestBody }), status: response.status,
      responseHeaders, responseBody, required: true };
    try { fs.appendFileSync(networkFile, JSON.stringify(exchange) + "\n", { mode: 0o600 }); } catch {}
    return response;
  };
  return () => { (globalThis as any).fetch = original; };
}

File: sdk/reproit-tui-py/reproit_tui_py/causal.py
"""Runner-only urllib capture/replay using PTY-safe side files."""

import io
import json
import os
import re
import urllib.request
import urllib.parse

_SECRET = re.compile(
    r"password|passwd|secret|token|authorization|cookie|email|phone|"
    r"api[-_. ]?key|publishable[-_. ]?key|private[-_. ]?key|"
    r"access[-_. ]?key|signing[-_. ]?key",
    re.I,
)


def _redact(value):
    if isinstance(value, list):
        return [_redact(v) for v in value]
    if isinstance(value, dict):
        out = {}
        for key in sorted(value):
            child = value[key]
            if _SECRET.search(str(key)):
                out[str(key)] = (
                    "<reproit:string:length=%d>" % len(child)
                    if isinstance(child, str)
                    else "<reproit:secret>"
                )
            else:
                out[str(key)] = _redact(child)
        return out
    return value


def _canonical(raw):
    parsed = urllib.parse.urlsplit(raw)
    query = urllib.parse.urlencode(
        sorted(urllib.parse.parse_qsl(parsed.query, keep_blank_values=True))
    )
    return urllib.parse.urlunsplit(
        (parsed.scheme, parsed.netloc, parsed.path, query, parsed.fragment)
    )


class _MemoryResponse:
    def __init__(self, body, status, headers, url):
        self._body = io.BytesIO(body)
        self.status = status
        self.code = status
        self.headers = headers
        self.url = url

    def read(self, amt=-1):
        return self._body.read(amt)

    def getcode(self):
        return self.status

    def geturl(self):
        return self.url

    def close(self):
        self._body.close()

    def __enter__(self):
        return self

    def __exit__(self, *_args):
        self.close()


def install_causal_urllib(exclude_prefix=None):
    network = os.environ.get("REPROIT_NETWORK_FILE")
    capsule_path = os.environ.get("REPROIT_CAPSULE")
    if not network and not capsule_path:
        return lambda: None
    original = urllib.request.urlopen
    capsule = None
    if capsule_path:
        try:
            with open(capsule_path, "r", encoding="utf-8") as handle:
                capsule = json.load(handle)
        except Exception:
            capsule = {"exchanges": []}
    exchanges = (capsule or {}).get("exchanges", [])
    used = set()
    prior_action = [0]
    ordinal = [0]

    def action_index():
        try:
            with open(os.environ.get("REPROIT_ACTION_FILE", ""), "r", encoding="utf-8") as handle:
                return int(handle.read().strip() or "0")
        except Exception:
            return 0

    def wrapped(request, *args, **kwargs):
        req = (
            request
            if isinstance(request, urllib.request.Request)
            else urllib.request.Request(request)
        )
        url = req.full_url
        if exclude_prefix and url.startswith(exclude_prefix):
            return original(request, *args, **kwargs)
        action = action_index()
        if action != prior_action[0]:
            prior_action[0] = action
            ordinal[0] = 0
        this_ordinal = ordinal[0]
        ordinal[0] += 1
        method = req.get_method().upper()
        actor = os.environ.get("REPROIT_DEVICE", "a")
        if capsule is not None:
            for index, exchange in enumerate(exchanges):
                if (
                    index not in used
                    and exchange.get("required")
                    and exchange.get("actor") == actor
                    and exchange.get("actionIndex") == action
                    and str(exchange.get("method", "")).upper() == method
                    and _canonical(str(exchange.get("url", ""))) == _canonical(url)
                ):
                    used.add(index)
                    body = exchange.get("responseBody", "")
                    raw = body.encode() if isinstance(body, str) else json.dumps(body).encode()
                    return _MemoryResponse(
                        raw,
                        int(exchange.get("status", 200)),
                        exchange.get("responseHeaders", {}),
                        url,
                    )
            raise RuntimeError("CAPSULE:MISS %s %s action=%d" % (method, url, action))

        response = original(request, *args, **kwargs)
        raw = response.read()
        headers = dict(response.headers.items())
        try:
            response_body = (
                _redact(json.loads(raw.decode()))
                if "json" in headers.get("Content-Type", headers.get("content-type", ""))
                else "<reproit:body:length=%d>" % len(raw)
            )
        except Exception:
            response_body = "<reproit:invalid-json>"
        request_body = None
        if req.data:
            try:
                request_body = _redact(json.loads(req.data.decode()))
            except Exception:
                request_body = "<reproit:body:length=%d>" % len(req.data)
        safe_headers = {
            k: ("<reproit:secret>" if _SECRET.search(k) else v) for k, v in req.header_items()
        }
        safe_response_headers = {
            k: ("<reproit:secret>" if _SECRET.search(k) else v) for k, v in headers.items()
        }
        exchange = {
            "id": "%s-%d-%d" % (actor, action, this_ordinal),
            "actor": actor,
            "actionIndex": action,
            "ordinal": this_ordinal,
            "protocol": urllib.parse.urlsplit(url).scheme,
            "method": method,
            "url": _canonical(url),
            "requestHeaders": safe_headers,
            "status": response.status,
            "responseHeaders": safe_response_headers,
            "responseBody": response_body,
            "required": True,
        }
        if request_body is not None:
            exchange["requestBody"] = request_body
        try:
            with open(network, "a", encoding="utf-8") as handle:
                handle.write(json.dumps(exchange, sort_keys=True, separators=(",", ":")) + "\n")
        except Exception:
            pass
        return _MemoryResponse(raw, response.status, headers, url)

    urllib.request.urlopen = wrapped
    _merge_capabilities(capsule is not None)
    return lambda: setattr(urllib.request, "urlopen", original)


def _merge_capabilities(_replay):
    path = os.environ.get("REPROIT_CAPABILITIES_FILE")
    if not path:
        return
    try:
        with open(path, "r", encoding="utf-8") as handle:
            value = json.load(handle)
    except Exception:
        value = {}
    value["http"] = {"status": "captured", "detail": "Python urllib"}
    value["http_replay"] = {"status": "captured", "detail": "Python urllib fail-closed replay"}
    try:
        with open(path, "w", encoding="utf-8") as handle:
            json.dump(value, handle, sort_keys=True, separators=(",", ":"))
    except Exception:
        pass
